---
title: "Genomic_Ranges_Assignment"
output: html_document
---
Setting up the files and importing key libraries
```{r}
setwd("C:/Users/Devin/Documents/GradSchool/IowaState/EEOB_546X/Genomic_Ranges_Assignment")
getwd()
source("http://bioconductor.org/biocLite.R")
biocLite()
biocLite("GenomicRanges")
library(BiocInstaller)
biocLite("GenomicFeatures")
library(IRanges)
library(GenomicRanges)
biocLite("rtracklayer")
library(rtracklayer)
```

Import transcript database of Mus musculus and chromosome 1. Sample psuedogene subtype as well and export as a GTF file.
```{r}
#mm_gtf<-import('C:/Users/Devin/Desktop/bds-files/chapter-09-working-with-range-data/Mus_musculus.GRCm38.75_chr1.gtf.gz')
biocLite("TxDb.Mmusculus.UCSC.mm10.ensGene")
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
txdb <- TxDb.Mmusculus.UCSC.mm10.ensGene
mm_gtf <- import('Mus_musculus.GRCm38.75_chr1.gtf.gz')
colnames(mcols(mm_gtf)) #lists all of the column names in mm_gtf
set.seed(0)
pseudogene_i <- which(mm_gtf$gene_biotype == "pseudogene" &
        mm_gtf$type == "gene")
pseudogene_sample <- sample(pseudogene_i, 5)
export(mm_gtf[pseudogene_sample], con="five_random_pseudogene.gtf",
      format="GTF")
```
Collapse the transcript db exons into non-overlapping ranges. Then find the exons from chromosome 1.
```{r}
collapsed_exons <- reduce(exons(txdb), ignore.strand=TRUE) #will reduce all of the various exons in txdb into non-overlapping ranges, save this as "collapsed_exons"
chr1_collapsed_exons <- collapsed_exons[seqnames(collapsed_exons) == "chr1"] # of collapsed_exons from txdb, find where the exons are from "chr1"
chr1_collapsed_exons
```

Import variant data from chromosome 1 and inspect it. 
```{r}
dbsnp137 <- import("mm10_snp137_chr1_trunc.bed.gz") #imports a file of variants from chromosome 1
summary(width(dbsnp137))
```
Adjust width to 1 in case the width is zero
```{r}
dbsnp137_resized <- dbsnp137
zw_i <- width(dbsnp137_resized) == 0 #boolean, finds out if the width of the variant is 0
dbsnp137_resized[zw_i] <- resize(dbsnp137_resized[zw_i], width=1) #resizes the width of the variants found by zw_i to 1
```
Find overlaps from resized variants with chromosome 1 exons and the proportion of those hits between the variants and chromosome 1 that are in an exon region.
```{r}
hits <- findOverlaps(dbsnp137_resized, chr1_collapsed_exons, 
        ignore.strand=TRUE) #chr1_collapsed_exons is the reduced non-overlapping ranges from chromosome 1

hits #there are 57624 hits!
length(dbsnp137_resized) #270000 variants in dbsnp137

length(unique(queryHits(hits))) #57623 hits are unique!
length(unique(queryHits(hits)))/length(dbsnp137_resized) #proportion of hits that are in chromosome 1 exons
```
Alternatively, finding the number of variants per exon can be done using the countOverlaps() function rather than using findOverlaps().Append the chr1_collapsed_exons with its number of variants per exon, column named "num_vars"
```{r}
 var_counts <- countOverlaps(chr1_collapsed_exons, dbsnp137_resized, ignore.strand=TRUE) #counts the number of overlaps/hits between the query and the sunject, the number is how many times the query hits
#var_counts
chr1_collapsed_exons$num_vars <- var_counts #appends the chr1_collapsed_exons num_variants column with var_counts (number of variants in exons)
write.csv(chr1_collapsed_exons, "chr1_collapsed_exons.csv", row.names = FALSE)
```
##Extra Credit
####Retrieving promoter regions:Flank VS Promoter
```{r}
table(mm_gtf$gene_biotype) #look at the mm_gtf subset gene_biotype to determine which type of genes we are interested in
chr1_pcg<-mm_gtf[mm_gtf$type == "gene" & mm_gtf$gene_biotype =="protein_coding"] #subset genes and pick out exclusively the protein coding genes 
summary(width(chr1_pcg))
length(chr1_pcg)
chr1_pcg_3kb_up<-flank(chr1_pcg,width=3000) #this will grab 3000 bp upstream from each feature to include the promoters
chr1_pcg_3kb_up
chr1_pcg_3kb_up2<-promoters(chr1_pcg,upstream=3000,downstream=0) #does the same thing as the code above! promoters() grabs 3000 upstream and 200 bp downstream by default.
identical(chr1_pcg_3kb_up,chr1_pcg_3kb_up2) #TRUE, the two are identical
```
####Retrieving protmoter Sequence:connection GenomicRanges with Sequence Data
Use the flank ranges to get the sequences of the promoter regions.
```{r}
biocLite("BSgenome")
biocLite("BSgenome.Mmusculus.UCSC.mm10")
library(BSgenome.Mmusculus.UCSC.mm10) #load the BSgenome.Mmusculus.UCSC.mm10
mm_gm<-BSgenome.Mmusculus.UCSC.mm10
organism(mm_gm) #verify the organiam 
providerVersion(mm_gm) #verify the provider version
provider(mm_gm) #verify the provider of the Msgenome.Mmusculus.UCSC.mm10
seqinfo(mm_gm)
```
Renaming the chromosome names manually.
```{r}
all(seqlevels(chr1_pcg_3kb_up)%in% seqlevels(mm_gm)) #check to see if the sequences we want are within the BSgenome object
gr<-GRanges(c("chr1","chr2"),IRanges(start=c(3,4),width=10)) #test GRange and manually change the name of a chromosome
seqlevels(gr)
seqlevels(gr)<-c("1","2") #change the names from chr1 and chr2 to 1 and 2
seqlevels(gr)



```
Renaming the chromosome names in chr1_pcg_3kb_up to match the chromosome names in Mus_musculus.GRCm38.75_chr.gft.gz. using convenient functions.
```{r}
seqlevelsStyle(chr1_pcg_3kb_up) #what style are the sublevels? 
seqlevelsStyle(mm_gm) #the two styles do not match!
seqlevelsStyle(chr1_pcg_3kb_up)<-"UCSC"#sets chr1_pcg_3kb_up to UCSC style
all(seqlevels(chr1_pcg_3kb_up)%in% seqlevels(mm_gm))

```
Grabbing sequences of promoter regions using GRanges
```{r}
chr1_3kb_seqs<-getSeq(mm_gm,chr1_pcg_3kb_up) #takes the sequences of mm_gm within the ranges given by the GRange object chr1_pcg_3kb_up
chr1_3kb_seqs
```
Write the promoter sequences to a FASTA file
```{r}
writeXStringSet(chr1_3kb_seqs,file="mm10_chr1_3kb_promoters.fasta")
```
